src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,764 @@
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_check_image_consistency_group>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(m_header_oid,
+                                          rados_completion, &op, &m_out_bl);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_check_image_consistency_group(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error fetching consistency group for image" << cpp_strerror(*result)
+                 << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  cls::rbd::GroupSpec s;
+  bufferlist::iterator it = m_out_bl.begin();
+  *result = librbd::cls_client::image_get_group_finish(&it, &s);
+  if (*result < 0) {
+    send_close_image(*result);
+    return nullptr;
+  }
+  if (s.is_valid()) {
+    lderr(m_cct) << "image is in a consistency group - not removing" << dendl;
+    send_close_image(-EMLINK);
+    return nullptr;
+  }
+
+  trim_image();
+
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::trim_image() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_trim_image>(this);
+
+  RWLock::RLocker owner_lock(m_image_ctx->owner_lock);
+  librbd::operation::TrimRequest<I> *req = librbd::operation::TrimRequest<I>::create(
+    *m_image_ctx, ctx, m_image_ctx->size, 0, m_prog_ctx);
+  req->send();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_trim_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "warning: failed to remove some object(s): " << cpp_strerror(*result)
+                 << dendl;
+  }
+
+  remove_child();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_child() {
+  ldout(m_cct, 20) << dendl;
+
+  m_image_ctx->parent_lock.get_read();
+  parent_info parent_info = m_image_ctx->parent_md;
+  m_image_ctx->parent_lock.put_read();
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::remove_child(&op, parent_info.spec, m_image_id);
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_remove_child>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_CHILDREN, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_remove_child(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result == -ENOENT) {
+    *result = 0;
+  } else if (*result < 0) {
+    lderr(m_cct) << "error removing child from children list" << cpp_strerror(*result)
+                 << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  if (!m_old_format) {
```

Comment:
Nit: this check could actually be moved to ```remove_child``` since by definition a v1 image cannot be a parent/clone